### PR TITLE
feature(views): add first and last page number to pagination

### LINKED ISSUE #5055: Improving site navigation via pagination links (Trac #5055)
_Original ticket http://trac.elgg.org/ticket/5055 on 43110044-02-08 by trac user iionly, assigned to unknown._

Elgg version: 1.8

There are two things that are slightly cumbersome when navigating through a site using the pagination links:

* the pagination links are only displayed at the bottom of a list. If I want to go to the next page (or apply any other offset), I would need to scroll down to the bottom first to be able to do so. If I can't reach the desired offset with one step I would need to repeat this (unnecessary scrolling down) maybe even several times.

* The pagination links are okay for applying a small offset in pagination. But I think there are a few important landing points missing. First, the links that show are:

"Previous", 1, 2, 3, 4, 5, 6, "Next"

What if I want to jump to the last page? What if I want to apply some larger offset to go through the list? I think the pagination links might work better with some additional options. For example:

"First", "-10", "Previous", 1, 2, 3, 4, 5, 6, "Next", "+10", "Last"

I'm not sure what would be the best way to present links to the direct neighbors of a page (how many links to offer for preceding and following pages). The previous and next pages respectively are already accessible by the "Previous" and "Next" links. So, there's kind of a redundance here. Maybe it's also possible to do it this way:

"First", "-10", "-5", "-2", "Previous", <Current page #>, "Next", "+2", "+5", "+10", "Last" | "<Goto page...>"

In this case <Current page #> would only display the page number of the currently displayed page. The last option "<Goto page...>" would allow to enter the desired page number directly.

### PR DESCRIPTION
**Pull request by jdalsem**
ref #5055 

pre-1.8 had this feature, but was removed for unknown reasons... this adds it back in a new way

before fix screenshot:
before

after fix screenshot:
after

### CODE REVIEW

**Comment by brettp** [comment]
Like it. One of my projects adds start and end links. I don't care for their language, but it might be a good addition, too? 
screen shot 2015-03-09 at 1 06 54 pm

**Comment by iionly** [comment]
While it's surely nice to have the "First"/"Last" links back (especially the "First" link), wouidn't it make sense to also include some jump options like +10/-10 (or +5/-5) as suggested in the ticket? It's kind of slow to navigate single page by single page if you know you want to jump to several pages back or forward (for example to get to the first new entry since your last visit).

**Comment by hypeJunction** [comment]
Perhaps just a text input field with a page number and next/prev and
last/first. Works nicely with responsive themes, otherwise all these
numbers with padding and margins span across multiple lines.

**Comment by jdalsem** [comment]
> start and end links

Thought about these too, but are they not a lot redundancy if you always see first and last page number also


> jump options like +10/-10 (or +5/-5) as suggested in the ticket

Don't really like this myself, but primarily this PR is first meant to restore previous pagination functionality, so we do not lose first and last indicators


> Perhaps just a text input field with a page number and next/prev and last/first.

Also a valid idea, but maybe for another ticket where we introduce multiple pagination types

**Comment by iionly** [comment]
If this PR is primarily meant to bring back the first/last links only it's not fully fixing issue https://github.com/Elgg/Elgg/issues/5055.

I think an easy way to jump over several pages backwards/forwards is quite useful (directly editing the offset in the url isn't really elegant...). Especially, it's annoying that you have to scroll down each time before being able to go to the next page if you really want to leap over sevaral pages to reach the desired page.

The idea of @hypeJunction with an input field could be an alternative to several separate links. Though I'm not sure if a dropdown selection with the available options might be the better choice over a text input.

File: views/default/navigation/pagination.php [hunk 1]
```diff
@@ -37,103 +37,118 @@
...
 
-	$pages->prev['href'] = elgg_http_add_url_query_elements($base_url, array($offset_key => $prev_offset));
+$pages = array();
 
-	$first_page = $current_page - $delta;
-	if ($first_page < 1) {
-		$first_page = 1;
-	}
+// determine starting page
+$start_page = max(min([sanitize_int($current_page - 2, true), sanitize_int($total_pages - 4, true)]), 1);
```
Comment (mrclay) [inline views/default/navigation/pagination.php]: Why sanitize_int here? doesn't seem to do anything.

**Comment by mrclay** [comment]
This PR seems clearly better than current behavior without any changes likely to affect JS/CSS, etc. Seems good for 1.11 and at the very least having a link to the 1st page seems essential.

It does not fix #5055 as that issue doesn't reach a consensus and is probably too broad in scope.

File: views/default/navigation/pagination.php (hunk 1 above)
Comment (jdalsem) [inline views/default/navigation/pagination.php]: it makes positive ints (with the true flag)

Comment (ewinslow) [inline views/default/navigation/pagination.php]: Ugh. We need Math::abs or something. Lol what a secretive behavior based on
a boolean...

Comment (mrclay) [inline views/default/navigation/pagination.php]: No, the false flag sets 0 as minimum bound, true just casts.

Comment (mrclay) [inline views/default/navigation/pagination.php]: $current_page and $total_pages are already ints. abs() will work.

**Comment by jdalsem** [comment]
changed the issue reference from fix to ref and removed the sanitise_int function in determining the startpage (as it was not needed at all)

**Comment by hypeJunction** [comment]
I think we should not hard code totals. Let's make it a setting instead of adding/substrating 4,5,6 etc.

**Comment by hypeJunction** [comment]
Never mind, I misread the code.

**Comment by hypeJunction** [comment]
I am thinking of adding a new page to theme sandbox to preview lists with varying number of pages. I need to it for #7912, but it will be good for PRs like this

**Comment by hypeJunction** [comment]
Did you mean to merge without more eyes on it. I didn't have to time to actually pull it in an look at the changes.
Having looked at the screenshot above, I figured I didn't actually misread the code. I think the number of pages between the spacers should be configurable. 5 is a very arbitrary number.

**Comment by jdalsem** [comment]
> 5 is a very arbitrary number.

it is indeed, but it matches the previous amount of pages (so width stays the same for example). Don't think a setting would be a good thing, but maybe fetching it from a $vars variable is an idea...

**Comment by hypeJunction** [comment]
By 'setting' I didn't mean a config or plugin setting value :) Yeah, something from $vars would work. E.g. for narrow containers (widgets) you may want to use less, for wider container - more

**Comment by jdalsem** [comment]
Feel free to provide a new PR for that. There are more things that could have been variable in this view (which i initially thought of), but that got out of hand very quick.

You could also think about:
 - show_page_numbers (show page numbers at all)
 - previous/next text (to switch to older/newer for example)
 - show previous/next (for ajax pagination where you only have older comments, like a show more)

**Comment by hypeJunction** [comment]
Ok. I am going to work on some list related stuff now, so I will see if I can squeeze it in somewhere. I guess I will have to do some rebasing now to accommodate my older PRs :(

**Comment by hypeJunction** [comment]
BTW, this looks beautiful. Thanks!

**Comment by michelesbw** [comment]
Ok tested, works.
But imho, this won't be liked a lot by sites' admins because especially in a community starting phase very few like to show immediately how few pages are there.
I think ajax (@hypeJuntion should have a plugin for this) is better accepted or at least a choice on which to use.

**Comment by jdalsem** [comment]
> But imho, this won't be liked a lot by sites' admins because especially in a community starting phase very few like to show immediately how few pages are there.

the previous functionality was that you only saw there were 6 pages of content (even if there were hundreds). That could make me think that there is not much happening... (so it is also true the other way around)

**Comment by michelesbw** [comment]
So maybe Brett's without "1464 total" would be better? (just saw it)

**Comment by jdalsem** [comment]
> So maybe Brett's without "1464 total" would be better? (just saw it)

doh... where were you a month ago :)

**Comment by michelesbw** [comment]
Sorry....from now on just "works for me" or "not" :)

**Pull request merged**